src/libfsm/minimise_test_oracle.c

```diff
@@ -0,0 +1,279 @@
+ */
+
+#include <stddef.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+#include <fsm/fsm.h>
+#include <fsm/capture.h>
+#include <fsm/pred.h>
+#include <fsm/walk.h>
+
+#include <adt/alloc.h>
+#include <adt/set.h>
+#include <adt/edgeset.h>
+#include <adt/stateset.h>
+#include <adt/hashset.h>
+#include <adt/u64bitset.h>
+
+#include <assert.h>
+
+#include "internal.h"
+#include "capture.h"
+
+#define LOG_NAIVE_MIN 0
+
+static fsm_state_t
+find_dst(const struct fsm *fsm,
+    fsm_state_t from, unsigned char label)
+{
+	const size_t state_count = fsm_countstates(fsm);
+	if (from == state_count) {
+		goto dead_state;
+	}
+	assert(from < state_count);
+
+	struct fsm_edge e;
+	if (edge_set_find(fsm->states[from].edges, label, &e)) {
+		return e.state;
+	}
+
+	/* edge not found for label */
+dead_state:
+	return state_count;	/* dead state */
+}
+
+struct fsm *
+fsm_minimise_test_oracle(const struct fsm *fsm)
+{
+	assert(fsm != NULL);
+	if (fsm_capture_has_captures(fsm)) {
+		assert(!"cannot be used with captures");
+		return NULL;
+	}
+
+	const size_t state_count = fsm_countstates(fsm);
+	{
+		fsm_state_t start;
+		if (state_count == 0 && !fsm_getstart(fsm, &start)) {
+			return fsm_clone(fsm);
+		}
+	}
+
+	struct fsm *res = NULL;
+
+	/* Fixpoint-based algorithm for calculating equivalent states
+	 * in a DFA, from pages 68-69 ("A minimization algorithm") of
+	 * _Introduction to Automata Theory, Languages, and Computation_
+	 * by Hopcroft & Ullman.
+	 *
+	 * Allocate an NxN bit table, used to track which states have
+	 * been proven distinguishable.
+	 *
+	 * For two states A and B, only use the half of the table where
+	 * A <= B, to avoid redundantly comparing B with A. In other
+	 * words, only the bits below the diagonal are actually used.
+	 *
+	 * An extra state is added to internally represent a complete
+	 * graph -- for any state that does not have a particular label
+	 * used elsewhere, treat it as if it has that label leading to
+	 * an extra dead state, which is non-accepting and only has
+	 * edges leading to itself.
+	 *
+	 * To start, mark all pairs of states in the table where one
+	 * state is an end state and the other is not. This indicates
+	 * that they are distinguishable by their result. Then, using
+	 * the Myhill-Nerode theorem, mark all pairs of states where
+	 * following the same label leads each to a pair of states that
+	 * have already been marked (and are thus transitively
+	 * distinguishable). Repeat until no furhter progress can be
+	 * made. Any pairs of states still unmarked can be combined into
+	 * a single state, because no possible input leads them to
+	 * observably distinguishable results.
+	 *
+	 * This method scales poorly as the number of states increases
+	 * (since it's comparing everything against everything else over
+	 * and over, for every label present), but it's straightforward
+	 * to implement and to check its results by hand, so it works
+	 * well as a test oracle. */
```

Comment:
I really appreciate this comment. Also I'm amazed that you can just like, write out an implementation of this. It took me weeks to do the first implementation of NFA->DFA conversion!

Reply:
To be fair, I wrote that in somewhat modified form about three years ago (https://github.com/katef/libfsm/blob/main/theft/fuzz_minimise.c#L306) -- this was adapting an existing thing, which was working with different input/output structures and not using ADTs added later on, rather than writing it from scratch. Looking at the commit logs, it seems like that took a couple hours initially.